src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -125,15 +125,15 @@ public override IEnumerable CreateDerivedEnumerableInstance(JsonPropertyInfo col
         {
             object instance = collectionPropertyInfo.DeclaredTypeClassInfo.CreateObject();
 
-            if (instance is IList instanceOfIList)
+            if (instance is IList instanceOfIList && !instanceOfIList.IsReadOnly)
```

Comment:
Wouldn't it be better to put this check inside the if block, e.g.
```C#
if (instance is IList instanceOfIList)
{
    if (!instanceOfIList.IsReadOnly)
    {
    }
}
```
and the same for the other case below?  If this instance is a readonly IList, then we don't need to continue processing the rest of the cascading else ifs, do we?

Reply:
> Wouldn't it be better to put this check inside the if block

Yes, it would, but you'd have to duplicate the throwing code below, which is probably fine. I can do a follow-up PR against `master` with this change. Do you think it's important enough to backport to this branch?

Reply:
> but you'd have to duplicate the throwing code below

Why?

> Do you think it's important enough to backport to this branch?

No

Reply:
Ah, looking at your code again, I see you suggested to only do the `Add` calls if `IsReadOnly` is `false`. That means read-only collections would just be ignored (empty) instead of throwing a `NotSupportedException`. That's why I mentioned the duplication of throwing code.

Reply:
> That means read-only collections would just be ignored (empty) instead of throwing a NotSupportedException

No.  We'd fall through and still hit the throw at the bottom of the method.  All the suggested change does is avoid executing the rest of the else ifs that won't be true anyway.

Reply:
Ah, DOH! Of course, those return statements would go inside that `if` as well 🤦

Reply:
Submitted https://github.com/dotnet/corefx/pull/40446. Good catch! 👍

Reply:
Submitted https://github.com/dotnet/corefx/pull/40446. Good catch! 👍